### ISSUE TITLE
File Organization

### ISSUE THREAD

**Issue by Doc-Maxx** [Mikemikenmike/effective-adventure#2]
We should establish a referable file organization structure and gain a better understanding of how Unity decides where to put new files. For example, script files pertaining to procedural generation could be located here:

- \Assets\scripts\ProceduralGeneration

I've encountered an issue referencing other codes already, although I may not be  familiar enough with c# to have it completely down, I think having everything code we write in an easy to follow structure should alleviate these problems.

This problem came up when adding a component script to an object. The file was placed into Assets folder, despite this specific script being attached to a prefab. Another problem I've encountered is an error from Unity when I attempt to drag a script from a folder and attempt to attach it to the item.

**Comment by JakeNeighbors** [comment]
@Praxian Placing code in folders using a proper folder hierarchy is a great idea. It would probably be good to put code under: `/Assets/src/`? Then have folders for `gui`, `server`, `procedural_generation/terrain`, `procedural_generation/trees`, `ai` and what not. It is typically not too bad to move the folders around (especially in C#).

However, this will not really solve the particular issue you address. C# could cares less what folder you place your code in or what you name your files. This is rather opposite of other languages that use the name of the file path to import and reference code from one file into another. C#'s compiler could care less. The only thing that really matters to the compiler is segregating your code into namespaces.

Unity3D actually covers this very problem in an article found here: [https://docs.unity3d.com/Manual/Namespaces.html](https://docs.unity3d.com/Manual/Namespaces.html). Without specifying a namespace, everything will be placed under the root namespace, You can only use references that are specified under that resource. To access references outside of the namespace you are in, you may simply specify the namespace with the `using` command. The `using` statement would be close to the `import` statement of other languages in that regard.

You may also specify a namespace over and over again in multiple files, No need to use the `using` statement on its own namespace either, even though the code you want to reference is from another file.

For additional structure, I would recommend laying the namespaces similar to the folder structure. Ie. the namespace for the file `src/gui/settings/graphics_settings.cs` might be `gui.settings` or perhaps just `gui`. It all depends on how you want to isolate code. Changing namespaces will, however, absolutely cause problems for code that imported the namespace using the old name. To avoid this, instead of simply renaming the namespace manually, please use JetBrain's refactor tool which will find every reference of the namespace in use and switch all references to the new name (right click old name -> refactor -> rename). The same refactoring technique may be used for variable names, class names, etc.

**Comment by Mikemikenmike** [comment]
@JakeNeighbors what would go inside an /src/ subfolder? Only scripts organized by more folders, or Prefabs, Materials, etc as well?

I was under the impression that the Assets folder was basically our src folder, so organization could simply be:

/Assets/Scripts
/Assets/Prefabs
/Assets/Materials

@Praxian 

> This problem came up when adding a component script to an object. The file was placed into Assets folder, despite this specific script being attached to a prefab.

This is unity's default behavior when you create a new script by adding it as a component in the editor. You would then have to manually drag it into the folder you want. Jake covered why that shouldn't result in errors. Your other option is to navigate to the folder you want to create the script in, right click, and create a new one directly.

**Comment by JakeNeighbors** [comment]
@Mikemikenmike I personally strongly dislike folders staring with capital letters. But the idea I had for `src` was that all the `Scripts` based items would go under `src`. Although, calling large source code components "scripts" is a bit off putting since scripts are meant to be small light-weight code.

Beyond just the source code material I don't think I would necessarily follow their model for "Prefabs", "Materials", "Textures" etc. It is an extremely frustrating concept to use. For example, given that you want to import a model into unity and modify as necessary, instead of placing all the components into one folder and have it all accessible right there, you essentially require yourself to separate those components needlessly into sub-folders making it difficult to track down components for the model. Unity themselves goes back and fourth on this where many developers ditch that structure entirely.

I would place generic code into something like `src` or `engine` (the `Script` equivalent). But literal "scripts" can be placed under specific folders if they are tailored for them (ie a specific script to animate an object). Alternatively, if there are generic textures and materials, I would put that into their generic folder name at root level. But largely, this is not a common occurrence. My concern with these concepts is that `Prefabs`, `Materials`, etc become a drop-zone for the everything.

**Comment by Doc-Maxx** [comment]
Yeah I agree with @JakeNeighbors , the default folder set up certainly encourages a sprinkling of the constituents of the same files across a bunch of different addresses. Okay, so we should probably decide on a way to organize these files. So maybe one of use should go ahead a create a semi-generic structure and we'll just iterate where appropriate.

Regarding scripting: Say two object behaviours require behaviours that are two special cases of the a more general mathematical structure. From the perspective of debugging and performance, would it be advantageous to have two different scripts, or one more general script that requires an additional input? I suppose my real question is: Are fewer but longer more general scripts better than shorter more specific scripts? Is there a significant cost to calling and searching for scripts?

I imagine, that no specific case will necessarily matter, from a performance perspective, but problems would begin to emerge once we reach some threshold.

**Comment by JakeNeighbors** [comment]
@Praxian

> Are fewer but longer more general scripts better than shorter more specific scripts?

For the most part, calling functions is incredibly fast. This only becomes a problem when you are performing an immense amount of calls (in the 10s or 100s of millions) per second. At which you can consider "flattening" the call tree. On occasion, the compiler will help you out on this end. In C++, you can use `inline` as a suggestion to the compiler to do this trick. I'm not sure if there is one for C#.

> Is there a significant cost to calling and searching for scripts?

As for searching for scripts: good news! Once the code is compiled, there is absolutely 0 searching. All the searching is done by the linker (a stage in the compilation process that finds all the projects symbols, ie classes, variables, etc, and creates a mapping strategy). The compiler also does some optimization where necessary. I should also note that names (class names, variable names etc) are reduced to generic references. So feel free to make names as small and as large as you want for the compiler. However, for human readability, please make them sane.

In terms of software engineering and project structure: Unless they are tightly coupled place classes & such into their own file. Unity already somewhat forces this already when making a component. If there are a bunch of similar small size objects too small for separate classes, I would make an exception to put them into a single file. Make your best judgement I guess.

The only large consideration for us is to organize files in a way so they are easy to find as non-file-indexing humans. Also is kind of handy to yank a folder with all the contents necessary for it to run without scavenging in the deep for all its fragments.
